### PR TITLE
Smart Processor + Unified Multimodal Embedding Support + HF Tokenization

### PR DESCRIPTION
**Pull request by Rian354**
## TLDR
This PR adds HuggingFace tokenizer support to `TupleTimeTextProcessor` so clinical text flows through any existing model (`MLP`, `RNN`, `Transformer`) with **0** model side changes, declared via a single registry alias in input_schema. It also has a `TemporalFeatureProcessor` + `UnifiedMultimodalEmbeddingModel` that gives all temporal modalities (text, codes, images, vitals) a common dict based output interface, and supports cross modal temporal alignment via sorting events on a shared timeline, and outputs a universal `collate_temporal` collator/aggregator to address serialization issues and model specific custom collation/aggregation logic.

---

## Summary

This PR adds 2 features to PyHealth:

1. **Smart Processor** (`TupleTimeTextProcessor` + HuggingFace tokenizer), clinical text paired with timestamps flows through any existing model (`MLP`, `Transformer`, `RNN`, `MultimodalRNN`) with zero model-side changes.
2.`TemporalFeatureProcessor` + `UnifiedMultimodalEmbeddingModel`, a foundation for *all* temporal modalities (text, codes, images, numerics/vitals), to fix serialization issues and N×M model-processor coupling.

All changes are **fully backward compatible**.

---

### Before this PR

| Problem | Example |
|---|---|
| Manual tokenization outside the dataset pipeline | `AutoTokenizer` + custom collate per model |
| Processors returned incompatible tuple arities | `(Tensor, Tensor)` vs `(Tensor, Tensor, str)` vs `(List[str], Tensor, str)` |
| `List[str]` inside tuples -> litdata serialization broken | `InMemorySampleDataset` pickle workaround only |
| No processor declared its modality | Manual `if type_tag == "image"` checks scattered in models |
| `TimeseriesProcessor` discarded timestamps silently | No way to use timestamps for temporal fusion |

### After this PR

```python
from pyhealth.datasets import create_sample_dataset
from pyhealth.models import RNN  # or MLP, Transformer, MultimodalRNN

dataset = create_sample_dataset(
    samples=patient_samples,
    input_schema={
        "conditions": "sequence",
        "notes": (
            "tuple_time_text",                      # registry alias , no import needed
            {"tokenizer_model": "emilyalsentzer/Bio_ClinicalBERT", "max_length": 128},
        ),
    },
    output_schema={"label": "binary"},
)
model = RNN(dataset, embedding_dim=128)  # done , BERT auto-loaded
```

For full multimodal temporal fusion across text + codes + images + vitals:

```python
from pyhealth.models import UnifiedMultimodalEmbeddingModel
from pyhealth.datasets import collate_temporal
from torch.utils.data import DataLoader

model = UnifiedMultimodalEmbeddingModel(
    processors=dataset.input_processors,   # all must be TemporalFeatureProcessor
    embedding_dim=128,
)
loader = DataLoader(dataset, batch_size=8, collate_fn=collate_temporal)
# out["sequence"]: (B, S_total, 128) , all events sorted by time
# out["mask"]:     (B, S_total)
```

---

#### `pyhealth/processors/tuple_time_text_processor.py`

- New `tokenizer_model` init argument (optional str, HuggingFace model name/path)
- `process()` returns a **5-element tuple** when tokenized:
  `(input_ids, attention_mask, token_type_ids, time_tensor, type_tag)`
  and the original **3-element tuple** when not tokenized (backward compatable)
- `schema()` updated: tokenized -> `("value", "mask", "token_type_ids", "time", "type_tag")`, raw -> `("text", "time", "type_tag")`
- `is_token()` returns `True` iff tokenizer loaded
- `size()` exposes `tokenizer.vocab_size`
- Graceful fallback for models without `token_type_ids` (e.g. RoBERTa)

#### `pyhealth/models/embedding.py`

- Detects `is_token() == True`, auto-loads `AutoModel.from_pretrained(...)` as backbone
- Adds projection layer when `hidden_size != embedding_dim`
- `forward()` gains optional `masks` kwarg threaded to transformer's `attention_mask`
- **3D handling**: `(B, N, L)` -> flatten -> transformer -> CLS-pool -> unflatten to `(B, N, H)`
- Returns `(embedded, masks)` when `output_mask=True`; plain dict otherwise

#### `pyhealth/models/mlp.py`, `transformer.py`, `rnn.py`

- Schema-aware tuple unpacking using `processor.schema()`
- Mask propagation to `EmbeddingModel`
- **RNN 3D mask fix**: token-level `(B, N, L)` mask reduced to note-level `(B, N)` via `.any(dim=-1)` before `pack_padded_sequence` , prevents `IndexError`

#### `examples/smart_processor_clinical_text_tutorial.ipynb` *(new)*

End-to-end notebook: raw processor -> tokenized -> `EmbeddingModel` with 3D inputs -> `MLP`, `Transformer`, `RNN`, `MultimodalRNN`, gradient flow through BERT-tiny.

#### `tests/test_tuple_time_text_tokenizer.py` *(new, 4 tests)*

1. Raw (no tokenizer) mode , output schema and dtypes
2. Tokenized mode , shape, dtype, schema
3. Smoke test: `fit()` + `process()` do not raise
4. **Canonical form**: `("tuple_time_text", kwargs)` in `input_schema` end-to-end (dataset -> DataLoader -> MLP forward)

---

#### `pyhealth/processors/base_processor.py`

- **`ModalityType(str, Enum)`**: `CODE | TEXT | IMAGE | NUMERIC | AUDIO | SIGNAL`
- **`TemporalFeatureProcessor(FeatureProcessor)`** ABC , requires:
  - `modality() -> ModalityType`
  - `value_dim() -> int` , raw value size before learned embedding
  - `process(value) -> dict[str, Tensor]` , keys `"value"`, `"time"`, optional `"mask"`
  - Default `schema()` returns `("value", "time")`

#### Existing temporal processors updated (all backward-compatible)

| Processor | Now inherits | `modality()` | `value_dim()` | `process_temporal()` added |
|---|---|---|---|---|
| `StageNetProcessor` | `TemporalFeatureProcessor` | `CODE` | `len(vocab)` | `{"value", "time"}` |
| `StageNetTensorProcessor` | `TemporalFeatureProcessor` | `NUMERIC` | `n_features` | `{"value", "time"}` |
| `TupleTimeTextProcessor` | `TemporalFeatureProcessor` | `TEXT` | `vocab_size` | `{"value", "mask", "time"}` |
| `TimeImageProcessor` | `TemporalFeatureProcessor` | `IMAGE` | `C*H*W` | `{"value", "time"}` |

> `process()` tuple output is **unchanged** for all 4, the current existing models are unaffected.

#### `pyhealth/processors/temporal_timeseries_processor.py` *(new)*

`TemporalTimeseriesProcessor` , same uniform resampling + forward-fill as `TimeseriesProcessor`, but returns `{"value": (S, F), "time": (S,)}` instead of a bare tensor. Timestamps are preserved for temporal alignment.

#### `pyhealth/datasets/collate.py` *(new)*

`collate_temporal()` , universal DataLoader collator:
- `dict` values -> recurse and stack/pad each sub-key
- `Tensor` -> stack if same shape, `pad_sequence` if variable length
- `int/float` -> `torch.tensor`
- Other -> plain list

Replaces all per-model custom collate functions.

#### `pyhealth/models/unified_embedding.py` *(new)*

`SinusoidalTimeEmbedding` + `UnifiedMultimodalEmbeddingModel`:
- Rejects non-`TemporalFeatureProcessor` inputs early with clear error
- Modality-specific encoders: `nn.Embedding` (codes), `AutoModel` (text+projection), ResNet-18 (images), `nn.Linear` (numerics)
- Temporal alignment: cat all events -> sort by `time` -> + `SinusoidalTimeEmbedding` + learned `type_embedding`
- Returns `{"sequence": (B, S_total, E'), "time", "mask", "type_ids"}`

#### `tests/test_unified_multimodal.py` *(new, 18 tests)*

Covers: `ModalityType`, `isinstance` checks, `process_temporal()` for all 4 processors, `TemporalTimeseriesProcessor` (basic, fit, imputation), `collate_temporal` (fixed-length, variable-length), `SinusoidalTimeEmbedding`, `UnifiedMultimodalEmbeddingModel` (forward, rejection, gradient flow, time-sort correctness).

---

### Before
```python
# User had to tokenize manually outside the pipeline
from transformers import AutoTokenizer
tokenizer = AutoTokenizer.from_pretrained("bert-base-uncased")
# ... custom collate, manual mask handling, bespoke model code ...
```

### After (canonical form , so no direct processor import needed)
```python
from pyhealth.datasets import create_sample_dataset
from pyhealth.models import MLP  # or RNN, Transformer, MultimodalRNN

dataset = create_sample_dataset(
    samples=patient_samples,
    input_schema={
        "conditions": "sequence",
        "notes": (
            "tuple_time_text",                          # registry alias
            {
                "tokenizer_model": "emilyalsentzer/Bio_ClinicalBERT",
                "max_length": 128,
            },
        ),
    },
    output_schema={"label": "binary"},
)
model = MLP(dataset, embedding_dim=128, hidden_dim=256)
# EmbeddingModel detects is_token()==True and auto-loads the BERT encoder
```

Direct instantiation is also present:

```python
from pyhealth.processors import TupleTimeTextProcessor

proc = TupleTimeTextProcessor(
    tokenizer_model="emilyalsentzer/Bio_ClinicalBERT",
    max_length=128,
)
# proc.schema()    -> ("value", "mask", "token_type_ids", "time", "type_tag")
# proc.is_token()  -> True
# proc.modality()  -> ModalityType.TEXT
# proc.value_dim() -> 28996  (tokenizer vocab_size)
# proc.size()      -> <vocab_size>  (same, model-dependent)

input_ids, mask, type_ids, time, tag = proc.process((["Chest pain", "Discharge"], [0., 72.]))
# input_ids: Tensor[2, 128] int64  , one row per note
# mask:      Tensor[2, 128] int64
# type_ids:  Tensor[2, 128] int64
# time:      Tensor[2]      float32
# tag:       "note"         str
```

---

## Backward Compatibility

| Scenario | Behavior |
|---|---|
| `TupleTimeTextProcessor()` (no tokenizer) | Identical to pre-PR , returns `(List[str], Tensor, str)` |
| `StageNetProcessor.process()` | Unchanged , returns `(time_tensor, value_tensor)` tuple |
| `TimeImageProcessor.process()` | Unchanged , returns `(images, timestamps, "image")` tuple |
| Existing models (`MLP`, `RNN`, `Transformer`, `MultimodalRNN`) | Completely unaffected |
| `EmbeddingModel` without `masks` kwarg | `masks` defaults to `None`, identical behavior |
| Non-temporal processors (`SequenceProcessor`, `MultiHotProcessor`, etc.) | Fully unaffected , `TemporalFeatureProcessor` is opt-in |

---

## Testing

```bash
# Tuple-time-text tests (4 pass, ~7s with bert-tiny)
TOKENIZERS_PARALLELISM=false pytest tests/test_tuple_time_text_tokenizer.py -v

# Results:
# test_tuple_time_text_processor_no_tokenizer        PASSED  (backward compat)
# test_tuple_time_text_processor_with_tokenizer      PASSED  (tokenized processor API)
# test_tokenizer_integration_in_pyhealth_workflow    PASSED  (fit+process smoke test)
# test_tuple_in_schema_canonical_form                PASSED  (registry alias + kwargs in input_schema)
# 4 passed in 7.08s

# Unified multimodal tests (18 pass, ~4.5s, no downloads needed)
pytest tests/test_unified_multimodal.py -v

# Results:
# test_modality_type_values                         PASSED
# test_stagenet_is_temporal                         PASSED
# test_stagenet_tensor_is_temporal                  PASSED
# test_tuple_time_text_is_temporal                  PASSED
# test_time_image_is_temporal                       PASSED
# test_stagenet_process_temporal                    PASSED
# test_stagenet_tensor_process_temporal             PASSED
# test_temporal_timeseries_basic                    PASSED
# test_temporal_timeseries_fit                      PASSED
# test_temporal_timeseries_imputation               PASSED
# test_collate_temporal_basic                       PASSED
# test_collate_temporal_variable_length             PASSED
# test_sinusoidal_time_embedding_shape              PASSED
# test_sinusoidal_different_times_differ            PASSED
# test_unified_model_code_only                      PASSED
# test_unified_model_rejects_non_temporal           PASSED
# test_unified_model_gradient_flow                  PASSED
# test_unified_model_time_sort                      PASSED
# 18 passed in 4.48s
```

End-to-end gradient flow verified for all four models (`MLP`, `Transformer`, `RNN`, `MultimodalRNN`) using `prajjwal1/bert-tiny` as the encoder. `UnifiedMultimodalEmbeddingModel` gradient flow verified through BERT CLS -> projection -> time embedding -> type embedding.

---

## What This PR Didn't Change
Included as this PR is a bit long

- `FeatureProcessor` base class API (`is_token`, `schema`, `dim`, `spatial`) , all existing processors unmodified
- `StageNetProcessor.process()`, `TimeImageProcessor.process()` tuple outputs , untouched
- `StageNet`, `ConCare`, `RETAIN`, `StageAttentionNet` model implementations , no changes
- `InMemorySampleDataset` / `SampleBuilder` collation , unchanged; `collate_temporal` is a new opt-in collator
- litdata disk-backed serialization , tokenized tensors are litdata-compatible; the raw `List[str]` variant (no tokenizer) should use `InMemorySampleDataset` or pre-tokenize before writing to disk
---

### CODE REVIEW

**Review by jhnwu3 — COMMENTED** [review]
Some quick thoughts. I've been thinking we may need to create pyhealth/models/embedding/ and just make a standardized EmbeddingModel module, mostly to standardize how we're wrapping these details. I think this has been overdue for quite sometime @Logiquo and I have been running into lots of issues when these aren't defined.

Especially since I think our pipeline is definitely more of:

pyhealth.datasets -> pyhealth.tasks -> pyhealth.processors -> pyhealth.models.embeddings -> pyhealth.models in some sense. 

There's a lot of ways to embed something, and I think while we have a pretty good idea of what's useful in a first start. It's definitely something that in practice we can definitely iterate on. 


I really like the new processor abstractions. This is a good way of going about this!
